File: data-structures/stack/implementations/stack.ts
import { Stack } from "../interfaces/stack";

export class MyStack<T> implements Stack<T> {
  private readonly initialCapacity: number;
  private container: Array<T>
  private currentLength: number
  private currentCapacity: number
  private head: number

  public constructor(initialCapacity = 4) {
    if (initialCapacity <= 0) {
      throw new Error("initialCapacity must be greater than zero");
    }

    this.initialCapacity = initialCapacity;
    this.container = new Array(initialCapacity)
    this.currentLength = 0;
    this.head = 0
    this.currentCapacity = initialCapacity
  }

  public length(): number {
    return this.currentLength
  }

  public capacity(): number {
    return this.currentCapacity
  }

  public isEmpty(): boolean {
    return this.length() <= 0;
  }

  public peek(): T | undefined {
    return this.container[this.head-1]
  }

  public push(_value: T): void {
    this.ensureCapacity()
    this.container[this.head] = _value;
    this.head++
    this.currentLength++
  }

  public pop(): T | undefined {
    if(this.isEmpty()) return
    const val = this.container[this.head - 1]
    delete this.container[this.head - 1]
    this.head--
    this.currentLength--

    return val
  }

  public clear(): void {
    this.container = []
    this.currentLength = 0;
    this.head = 0
    this.currentCapacity = this.initialCapacity
  }

  public toArray(): T[] {
    const res: T[] = []

    for(let i = 0; i < this.length(); i++) {
      res.push(this.container[i])
    }

    return res
  }

  protected getInitialCapacity(): number {
    return this.initialCapacity;
  }

  private ensureCapacity() {
    if(this.length() < this.capacity()) return

    const newCapacity = this.currentCapacity * 2;
    const newContainer = new Array(newCapacity)

    for(let i = this.head-1; i >= 0; i--) {
      newContainer[i] = this.container[i]
    }

    this.container = newContainer
    this.currentCapacity = newCapacity


  }
}

File: data-structures/stack/implementations/stack.go
package implementations

import "dsa/data-structures/stack/interfaces"

type MyStack[T any] struct {
	initialCapacity int
	currentCapacity int
	currentLength   int
	container       []T
}

var _ interfaces.Stack[int] = (*MyStack[int])(nil)

func NewMyStack[T any](initialCapacity int) *MyStack[T] {
	if initialCapacity <= 0 {
		panic("initialCapacity must be greater than zero")
	}

	return &MyStack[T]{
		initialCapacity: initialCapacity,
		currentCapacity: initialCapacity,
		currentLength:   0,
		container:       make([]T, initialCapacity),
	}
}

func (s *MyStack[T]) Length() int {
	return s.currentLength
}

func (s *MyStack[T]) Capacity() int {
	return s.currentCapacity
}

func (s *MyStack[T]) IsEmpty() bool {
	return s.Length() <= 0
}

func (s *MyStack[T]) Peek() T {
	if s.IsEmpty() {
		panic("peek: stack is empty")
	}

	return s.container[s.currentLength-1]
}

func (s *MyStack[T]) Push(value T) {
	s.ensureCapacity()
	s.container[s.currentLength] = value
	s.currentLength++
}

func (s *MyStack[T]) Pop() T {
	if s.IsEmpty() {
		panic("pop: stack is empty")
	}

	val := s.container[s.currentLength-1]

	s.currentLength--
	return val
}

func (s *MyStack[T]) Clear() {
	s.currentLength = 0
	s.currentCapacity = s.initialCapacity
	s.container = make([]T, 0, s.initialCapacity)
}

func (s *MyStack[T]) ToSlice() []T {
	n := s.Length()
	res := make([]T, n)

	for i := 0; i < n; i++ {
		res[i] = s.container[i]
	}

	return res
}

func (s *MyStack[T]) InitialCapacity() int {
	return s.initialCapacity
}

func (s *MyStack[T]) ensureCapacity() {
	if s.Length() < s.Capacity() {
		return
	}

	capacity := s.currentCapacity * 2
	container := make([]T, capacity)

	for i := 0; i < s.Length(); i++ {
		container[i] = s.container[i]
	}

	s.currentCapacity = capacity
	s.container = container
}
